File: python/130. Surrounded Regions.py
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if not board:
            return
        m, n = len(board), len(board[0])
        queue = []
        for i in range(m):
            for j in range(n):
                if (i == 0 or j == 0 or i == m-1 or j == n-1) and board[i][j] == "O":
                    queue.append((i,j))
        while queue:
            x, y = queue.pop(0)
            board[x][y] = "V"
            if x -1 >= 0 and board[x-1][y] == "O":
                queue.append((x-1, y))
            if x + 1 < m and board[x+1][y] == "O":
                queue.append((x+1, y))
            if y - 1 >= 0 and board[x][y-1] == "O":
                queue.append((x, y - 1))
            if y + 1 < n  and board[x][y+1] == "O":
                queue.append((x, y + 1))
        for i in range(m):
            for j in range(n):
                if board[i][j] == "O":
                    board[i][j] = "X"
                elif board[i][j] == "V":
                    board[i][j] = "O"
        

File: cpp/101-150/130. Surrounded Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        int i, j;
        int row = board.size();
        if (row == 0)
            return;
        int col = board[0].size();
        for (i = 0; i < row; i++) {
            check(board, i, 0, row, col);
            if(col > 1)
            check(board, i, col - 1, row, col);
        }
        for (j = 0; j < col; j++) {
            check(board, 0, j, row, col);
            if(row > 1)
            check(board, row - 1, j, col, row);
        }
        for (i = 0; i < row; i++) {
            for (j = 0; j < col; j++) {
                board[i][j] = (board[i][j] == '1') ? 'O' : 'X';
            }
        }
    }
	void check(vector<vector<char> >&vec,int i,int j,int row,int col){
		if(vec[i][j]=='O'){
			vec[i][j]='1';
			if(i>1)
				check(vec,i-1,j,row,col);
			if(j>1)
				check(vec,i,j-1,row,col);
			if(i+1<row)
				check(vec,i+1,j,row,col);
			if(j+1<col)
				check(vec,i,j+1,row,col);
		}
	}
};